kymatio/scattering1d/filter_bank.py

```diff
@@ -351,125 +333,59 @@ def scattering_filter_factory(N, J, Q, T, r_psi=math.sqrt(0.5),
-    phi_f = {}
-    psi1_f = []
-    psi2_f = []
-
-    # compute the band-pass filters of the second order,
-    # which can take as input a subsampled
-    for (xi2, sigma2, j2) in zip(xi2s, sigma2s, j2s):
-        # compute the current value for the max_subsampling,
-        # which depends on the input it can accept.
-        if max_subsampling is None:
-            possible_subsamplings_after_order1 = [j1 for j1 in j1s if j2 > j1]
-            if len(possible_subsamplings_after_order1) > 0:
-                max_sub_psi2 = max(possible_subsamplings_after_order1)
-            else:
-                max_sub_psi2 = 0
-        else:
-            max_sub_psi2 = max_subsampling
-        # We first compute the filter without subsampling
-
-        psi_levels = [morlet_1d(N, xi2, sigma2)]
-        # compute the filter after subsampling at all other subsamplings
-        # which might be received by the network, based on this first filter
-        for level in range(1, max_sub_psi2 + 1):
-            nperiods = 2**level
-            psi_levels.append(periodize_filter_fourier(psi_levels[0], nperiods))
-        psi2_f.append({'levels': psi_levels, 'xi': xi2, 'sigma': sigma2, 'j': j2})
-
-    # for the 1st order filters, the input is not subsampled so we
-    # can only compute them with N=2**J_support
-    for (xi1, sigma1, j1) in zip(xi1s, sigma1s, j1s):
-        psi_levels = [morlet_1d(N, xi1, sigma1)]
-        psi1_f.append({'levels': psi_levels, 'xi': xi1, 'sigma': sigma1, 'j': j1})
-
-    # compute the low-pass filters phi
-    # Determine the maximal subsampling for phi, which depends on the
-    # input it can accept (both 1st and 2nd order)
+    filterbank_fn, filterbank_kwargs = filterbank
+    max_j = 0
+    previous_J = 0
     log2_T = math.floor(math.log2(T))
-    if max_subsampling is None:
-        max_subsampling_after_psi1 = max(j1s)
-        max_subsampling_after_psi2 = max(j2s)
-        max_sub_phi = min(max(max_subsampling_after_psi1,
-                              max_subsampling_after_psi2), log2_T)
-    else:
-        max_sub_phi = max_subsampling
+    psis_f = []
+
+    # Loop over scattering layers
+    for Q_layer in Q:
+        psi_f = []
+
+        # Loop over center frequencies xi and bandwidths sigma in the filterbank
+        for xi, sigma in filterbank_fn(J, Q_layer, **filterbank_kwargs):
+
+            # Construct filter at full resolution
+            psi_levels = [morlet_1d(N, xi, sigma)]
+            j = get_max_dyadic_subsampling(xi, sigma, **filterbank_kwargs)
+
+            # Resample to smaller resolutions if necessary (beyond 1st layer)
+            # The idiom min(previous_J, j, log2_T) implements "j1 < j2"
+            for level in range(1, min(previous_J, j, log2_T)):
+                psi_level = psi_levels[0].reshape(2 ** level, -1).mean(axis=0)
+                psi_levels.append(psi_level)
+            psi_f.append({'levels': psi_levels, 'xi': xi, 'sigma': sigma, 'j': j})
+            max_j = max(j, max_j)
```

Comment:
or has it been moved here?

Reply:
the `filterbank_fn` yields `xi` and `sigma` in similar ways as `compute_params_filterbank` was returning them.

`get_max_dyadic_subsampling` is here (outside of the generator, see #856 ) because all it needs to know is `(xi, sigma, alpha)`.